libc/stdio_printf.c

```diff
@@ -0,0 +1,314 @@
+	TEST_IGNORE();
+
+	ftest("%1$d %2$d %3$d", 1 COMMA 2 COMMA 3, "3 2 1");
+}
+
+
+TEST_GROUP(sprintf_formatting);
+
+TEST_SETUP(sprintf_formatting)
+{ /* empty */
+}
+
+
+TEST_TEAR_DOWN(sprintf_formatting)
+{ /* empty */
+}
+
+
+TEST(sprintf_formatting, mod_width)
+{
+	/* check for space/zero padding and trucation */
+	stest("%5d", 123, "  123");
+	stest("%05d", 123, "00123");
+	stest("%5d", 1234567890, "1234567890");
+
+	stest("%10f", 1.23, "  1.230000");
+	stest("%010f", 1.23, "001.230000");
+
+	stest("%5s", "ab", "   ab");
+
+	/* width specified as additional argument */
+	stest("%*d", 5 COMMA 123, "  123");
+
+	TEST_IGNORE();
+	/* <posix incompliance> justyfication does not work for chars */
+	stest("%5c", 'a', "    a");
+
+	/* <posix incompliance> truncates string */
+	stest("%5s", "abcdefgh", "abcdefgh");
+}
+
+
+TEST(sprintf_formatting, mod_flags_ljust)
+{
+	stest("%-5s", "ab", "ab   ");
+	stest("%-*s", 5 COMMA "ab", "ab   ");
+
+	TEST_IGNORE();
+	/* <posix incompliance> left justyfication works only for strings */
+	stest("%-10f", 1.23, "1.230000  ");
+	stest("%-5c", 'a', "a    ");
+	stest("%-5d", 123, "123  ");
+
+	/* <posix incompliance> justyfication does not work for chars */
+	stest("%-5c", 'a', "a    ");
+}
+
+TEST(sprintf_formatting, mod_flags_sign)
+{
+	stest("%+d", -123, "-123");
+	stest("%+d", 123, "+123");
+	stest("%+f", -1.23, "-1.230000");
+
+	TEST_IGNORE();
+	/* <posix incompliance> sign modifier does not work for positive floats */
+	stest("%+f", 1.23, "+1.230000");
+}
+
+
+TEST(sprintf_formatting, mod_flags_hash)
+{
+	stest("%#o", 123, "0173");
+	stest("%#x", 123, "0x7b");
+}
+
+
+TEST(sprintf_formatting, mod_flag_precision)
+{
+	stest("%.5d", 12345678, "12345678");
+	stest("%.5o", 12345678, "57060516");
+	stest("%.5x", 12345678, "bc614e");
+
+	stest("%.3f", 1.2345678, "1.235");
+	stest("%.3f", 1.4, "1.400");
+
+	stest("%.3s", "a", "a");
+	stest("%.3s", "abcdefgh", "abc");
+
+	/* precision specified as additional argument */
+	stest("%.*f", 3 COMMA 1.2345678, "1.235");
+
+	TEST_IGNORE();
+	/* <posix incompliance> precision modifier does not work for shorter integers */
+	stest("%.5d", 123, "00123");
+	stest("%.5o", 123, "00173");
+	stest("%.5x", 123, "0007b");
+}
+
+
+TEST(sprintf_formatting, specifiers)
+{
+	stest("%ld", (long int)0x80000000, "-2147483648");
+	stest("%lu", (long unsigned int)0x80000000, "2147483648");
+
+	stest("%lld", (long long int)0x8000000000000000, "-9223372036854775808");
+	stest("%llu", (long long unsigned int)0x8000000000000000, "9223372036854775808");
+
+	stest("%jd", (intmax_t)(0x8000000000000000), "-9223372036854775808");
+	stest("%zd", (size_t)(0x80000000), "-2147483648");
+
+	TEST_IGNORE();
+	/* <posix incompliance>  prints format string */
+
+	stest("%hhd", 2147483520, "-128");
+	stest("%hhu", 2147483520, "128");
+
+	stest("%hd", 2147450880, "-32768");
+	stest("%hu", 2147450880, "32768");
+
+	stest("%Lf", (long double)1.0, "1.0");
```

Comment:
Maybe a value that really need a long double precision?

Reply:
Leaving as is. There is currently no way to test exponential printing (not supported), so it would need over 300 buffer size to  be tested with max value. Also we have issue with floats greater than 2^20 so untestable now.